### PR TITLE
[keymap] dactyl_left

### PR DESCRIPTION
**Pull request by RedForty**
Special layout for the left side of the ergodox dactyl.

## Description

Created a layout for the left side of the dactyl. I'm using it as a macro board.

### CODE REVIEW

File: keyboards/handwired/dactyl_left/readme.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# dactyl_left
+
+![dactyl_left](https://i.imgur.com/PYL3Ca2.png)
+
+A custom build of an ergodox dactyl using a Teensy. Pins are custom.
+
+Keyboard Maintainer: [RedForty](https://github.com/RedForty)  
+Hardware Supported: Left side of the dactyl only
```
Comment (noroadsleft) [inline keyboards/handwired/dactyl_left/readme.md]: This line needs to end in two spaces for GitHub to render it properly.

```suggestion
Hardware Supported: Left side of the dactyl only  
```

File: keyboards/handwired/dactyl_left/readme.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+Hardware Availability: Custom job
+
+Make example for this keyboard (after setting up your build environment):
+
+    make dactyl_left:default
```
Comment (noroadsleft) [inline keyboards/handwired/dactyl_left/readme.md]:
```suggestion
    make handwired/dactyl_left:default
```

File: keyboards/handwired/dactyl_left/rules.mk [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   atmega32a    bootloadHID
+BOOTLOADER = halfkay
+
+
+# If you don't know the bootloader type, then you can specify the
```
Comment (noroadsleft) [inline keyboards/handwired/dactyl_left/rules.mk]: Lines 15-24 should be deleted.

File: keyboards/handwired/dactyl_left/keymaps/default/layers.json [hunk 4]
```diff
@@ -0,0 +1 @@
..."KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_F6", "KC_F7", "KC_F8", "KC_F9", "KC_F10", "KC_F11", "KC_UP", "KC_7", "KC_8", "KC_9", "KC_ASTR", "KC_F12", "KC_DOWN", "KC_4", "KC_5", "KC_6", "KC_PLUS", "KC_TRNS", "KC_AMPR", "KC_1", "KC_2", "KC_3", "KC_BSLS", "KC_TRNS", "KC_TRNS", "KC_DOT", "KC_0", "KC_EQL", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS"], ["KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_MS_U", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_MS_L", "KC_MS_D", "KC_MS_R", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_BTN1", "KC_BTN2", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_MPLY", "KC_TRNS", "KC_TRNS", "KC_MPRV", "KC_MNXT", "KC_TRNS", "KC_TRNS", "KC_VOLU", "KC_VOLD", "KC_MUTE", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_WBAK"]]
```
Comment (noroadsleft) [inline keyboards/handwired/dactyl_left/keymaps/default/layers.json]: This file doesn't match your hardware. It should be removed from the pull request.

File: keyboards/handwired/dactyl_left/info.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+    "keyboard_name": "dactyl_left", 
+    "url": "", 
+    "maintainer": "RedForty", 
+    "width": 6.5,
```
Comment (noroadsleft) [inline keyboards/handwired/dactyl_left/info.json]:
```suggestion
    "width": 8, 
```

File: keyboards/handwired/dactyl_left/info.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
+    "keyboard_name": "dactyl_left", 
+    "url": "", 
+    "maintainer": "RedForty", 
+    "width": 6.5, 
+    "height": 7.375,
```
Comment (noroadsleft) [inline keyboards/handwired/dactyl_left/info.json]:
```suggestion
    "height": 7.75, 
```

File: keyboards/handwired/dactyl_left/info.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
+    "keyboard_name": "dactyl_left", 
+    "url": "", 
+    "maintainer": "RedForty", 
+    "width": 6.5, 
+    "height": 7.375, 
+    "layouts": {
+        "LAYOUT": {
+            "layout": [{"x":3.5, "y":0}, {"x":2.5, "y":0.125}, {"x":4.5, "y":0.125}, {"x":5.5, "y":0.25}, {"x":0, "y":0.375, "w":1.5}, {"x":1.5, "y":0.375}, {"x":3.5, "y":1}, {"x":2.5, "y":1.125}, {"x":4.5, "y":1.125}, {"x":5.5, "y":1.25}, {"x":0, "y":1.375, "w":1.5}, {"x":1.5, "y":1.375}, {"x":3.5, "y":2}, {"x":2.5, "y":2.125}, {"x":4.5, "y":2.125}, {"x":5.5, "y":2.25}, {"x":0, "y":2.375, "w":1.5}, {"x":1.5, "y":2.375}, {"x":3.5, "y":3}, {"x":2.5, "y":3.125}, {"x":4.5, "y":3.125}, {"x":5.5, "y":3.25}, {"x":0, "y":3.375, "w":1.5}, {"x":1.5, "y":3.375}, {"x":3.5, "y":4}, {"x":2.5, "y":4.125}, {"x":4.5, "y":4.125}, {"x":0.5, "y":4.375}, {"x":1.5, "y":4.375}, {"x":1, "y":4.375}, {"x":2, "y":4.375}, {"x":0, "y":5.375, "h":2}, {"x":1, "y":5.375, "h":2}, {"x":2, "y":5.375}, {"x":2, "y":6.375}]
```
Comment (noroadsleft) [inline keyboards/handwired/dactyl_left/info.json]: The vertical stagger of the columns means QMK's KLE to info.json converter doesn't order the key objects correctly (it counts `"y":0.125` as a different row from `"y":0`).

```suggestion
            "layout": [{"x":0, "y":0.375, "w":1.5}, {"x":1.5, "y":0.375}, {"x":2.5, "y":0.125}, {"x":3.5, "y":0}, {"x":4.5, "y":0.125}, {"x":5.5, "y":0.25}, {"x":0, "y":1.375, "w":1.5}, {"x":1.5, "y":1.375}, {"x":2.5, "y":1.125}, {"x":3.5, "y":1}, {"x":4.5, "y":1.125}, {"x":5.5, "y":1.25}, {"x":0, "y":2.375, "w":1.5}, {"x":1.5, "y":2.375}, {"x":2.5, "y":2.125}, {"x":3.5, "y":2}, {"x":4.5, "y":2.125}, {"x":5.5, "y":2.25}, {"x":0, "y":3.375, "w":1.5}, {"x":1.5, "y":3.375}, {"x":2.5, "y":3.125}, {"x":3.5, "y":3}, {"x":4.5, "y":3.125}, {"x":5.5, "y":3.25}, {"x":0.5, "y":4.375}, {"x":1.5, "y":4.375}, {"x":2.5, "y":4.125}, {"x":3.5, "y":4}, {"x":4.5, "y":4.125}, {"x":6, "y":4.75}, {"x":7, "y":4.75}, {"x":5, "y":5.75, "h":2}, {"x":6, "y":5.75, "h":2}, {"x":7, "y":5.75}, {"x":7, "y":6.75}]
```

File: keyboards/handwired/dactyl_left/rules.mk (hunk 3 above)
Comment (RedForty) [reply keyboards/handwired/dactyl_left/rules.mk]: Updated! Thank you!

File: keyboards/handwired/dactyl_left/keymaps/default/layers.json (hunk 4 above)
Comment (RedForty) [reply keyboards/handwired/dactyl_left/keymaps/default/layers.json]: Updated! Thank you!

**Comment by RedForty** [comment]
> Wasn't sure if this was right, but this is how QMK Configurator will assign your keys.
> 
> image
> 
> Please advise.

That looks about right as far as the layout goes. I want to use the Configurator to change what each key does, however. This is a good starting point.

**Comment by noroadsleft** [comment]
Are the keycodes on the intended keys, though?

**Comment by RedForty** [comment]
> Are the keycodes on the intended keys, though? Particularly the thumb cluster.

Eh, not really. I intend to change it entirely. 
Would it be helpful if the layout started as a copy of the Ergodox? That might make more sense.

**Comment by noroadsleft** [comment]
> Eh, not really. I intend to change it entirely.
> Would it be helpful if the layout started as a copy of the Ergodox? That might make more sense.

It might. I'm having a hard time working out which keycode in `keymap.c` goes on which physical key. The ["geographic" section of your layout macro](https://github.com/qmk/qmk_firmware/blob/fd196514f86f9106cb5b778d92a99c53e87e7082/keyboards/handwired/dactyl_left/dactyl_left.h#L28-L35) and your [keycode assignments](https://github.com/qmk/qmk_firmware/blob/fd196514f86f9106cb5b778d92a99c53e87e7082/keyboards/handwired/dactyl_left/keymaps/default/keymap.c#L22-L28) don't resemble the physical layout of the board.

**Comment by RedForty** [comment]
> > Eh, not really. I intend to change it entirely.
> > Would it be helpful if the layout started as a copy of the Ergodox? That might make more sense.
> 
> It might. I'm having a hard time working out which keycode in `keymap.c` goes on which physical key. The ["geographic" section of your layout macro](https://github.com/qmk/qmk_firmware/blob/fd196514f86f9106cb5b778d92a99c53e87e7082/keyboards/handwired/dactyl_left/dactyl_left.h#L28-L35) and your [keycode assignments](https://github.com/qmk/qmk_firmware/blob/fd196514f86f9106cb5b778d92a99c53e87e7082/keyboards/handwired/dactyl_left/keymaps/default/keymap.c#L22-L28) don't resemble the physical layout of the board.
> 
> This affects the ordering for the objects in info.json, so that you don't end up assigning a keycode to one key and then it's somewhere else in the flashed firmware running on the board.

Thank you for your help. Ok, I copied over the default keymaps for the ergodox from keyboard-layout-editor.com The left side, the bottom row, and the thumb cluster are all blank (space).

**Comment by RedForty** [comment]
Hey @noroadsleft , please let me know if there is anything else I need to do for this PR.

**Comment by noroadsleft** [comment]
@RedForty, sorry I missed this.

I still can't figure out which key in `keymap.c` is supposed to be which key in the Configurator implementation, but at the very least we could merge this in and fix it later.

**Comment by RedForty** [comment]
Ah, I think I understand your question now. The bottom row is the thumb cluster. But not in a smooth way (like working clockwise through the cluster). This is totally a handwired experiment. :)
I'm hoping to go ahead with the configurator and make the rest work.
Thank you!

**Comment by RedForty** [comment]
Hey @noroadsleft, is there anything else I should do for the merge? Thank you

**Comment by noroadsleft** [comment]
> Hey @noroadsleft, is there anything else I should do for the merge? Thank you

Can I blame catching a terrible cold and having felt largely brain-dead for the past week?

**Comment by noroadsleft** [comment]
Also, @RedForty, you should reset your master branch after this merge.

There are instructions on how to use your master branch with QMK here:

[**Best Practices: Your fork's master: Update Often, Commit Never**](https://docs.qmk.fm/#/newbs_best_practices?id=your-fork39s-master-update-often-commit-never)

**Comment by RedForty** [comment]
Thanks a ton! Hope you're feeling better!

**Pull request merged**